dask_expr/frameexpr.py

```diff
@@ -0,0 +1,1375 @@
+from __future__ import annotations
+
+import functools
+import numbers
+import operator
+from collections import defaultdict
+from collections.abc import Mapping
+
+import dask
+import pandas as pd
+from dask.base import normalize_token, tokenize
+from dask.dataframe import methods
+from dask.dataframe.core import (
+    _get_divisions_map_partitions,
+    _get_meta_map_partitions,
+    apply_and_enforce,
+    is_dataframe_like,
+    is_index_like,
+    is_series_like,
+    make_meta,
+)
+from dask.dataframe.dispatch import meta_nonempty
+from dask.utils import M, apply, funcname
+
+from dask_expr.expr import Expr
+
+replacement_rules = []
+
+no_default = "__no_default__"
+
+
+class FrameExpr(Expr):
+    """Primary class for all Expressions
+
+    This mostly includes pandas-like method
+    definitions to make us look more like a DataFrame.
+    """
+
+    def __hash__(self):
+        return hash(self._name)
+
+    @functools.cached_property
+    def ndim(self):
+        meta = self._meta
+        try:
+            return meta.ndim
+        except AttributeError:
+            return 0
+
+    def __getattr__(self, key):
+        try:
+            return object.__getattribute__(self, key)
+        except AttributeError as err:
+            # Allow operands to be accessed as attributes
+            # as long as the keys are not already reserved
+            # by existing methods/properties
+            _parameters = type(self)._parameters
+            if key in _parameters:
+                idx = _parameters.index(key)
+                return self.operands[idx]
+            if is_dataframe_like(self._meta) and key in self._meta.columns:
+                return self[key]
+            raise err
+
+    def _layer(self) -> dict:
+        """The graph layer added by this expression
+
+        Examples
+        --------
+        >>> class Add(FrameExpr):
+        ...     def _layer(self):
+        ...         return {
+        ...             (self._name, i): (operator.add, (self.left._name, i), (self.right._name, i))
+        ...             for i in range(self.npartitions)
+        ...         }
+
+        Returns
+        -------
+        layer: dict
+            The Dask task graph added by this expression
+
+        See Also
+        --------
+        Expr._task
+        Expr.__dask_graph__
+        """
+
+        return {(self._name, i): self._task(i) for i in range(self.npartitions)}
+
+    def _simplify_down(self):
+        return
+
+    def _simplify_up(self, parent):
+        return
+
+    def optimize(self, **kwargs):
+        return optimize(self, **kwargs)
```

Comment:
These should maybe be in `Expr`

Reply:
Oh, I see, optimize is dataframe specific. The simplify methods then at least should probably be moved up.

Reply:
The simplify steps live in both actually, missed deleting them here